CMake Build System

AlexMax opened the pull request:
So this commit is largely built off the prior work I had done for both DietBoom, and Chocolate Doom.  Thus far, I have confirmed that these work on Linux, Windows, and cross-compiling to Windows from Linux.  The windows executables are compiled properly and have the resource file and icon embedded.

I made some code changes to get the thing compiling with Visual Studio, as well as switching some of the automake-centric defines to their CMake equivalents.  I also spruced up the `convert-icon` script and added it as a distinct custom target.

I'm not sure if you consider this "done" or not - for one, it appears that automake had a `make dist` that the CMake replacement currently does not replicate.  Mac support is also untested.

As of this moment, I have left the existing automake files in place, but due to some defines changing they no longer actually work.  I merely leave them in place for personal reference, and once you are happy with the state of the CMake build system, I will remove them and add the removal commit to this pull request.

fabiangreffrath commented:
Wow, thank you very much! Looks like a massive effort.

What I am missing, though, is conditional defines for DOGS, BETA and RANGECHECK. While I will probably remove the first two in a future commit and have their code paths enabled by default, it would still be nice to be able to undefine RANGECHECK on purpose. Ah, and could I get a conditional define for INSTRUMENTED, defaulting to disabled, please?

AlexMax commented:
Made a couple of updates tonight.  Changed the name, added options, added the ability to gather and copy all DLL files to the binary directory automatically, that sort of thing.

I've also started teaching the build system how to generate a ZIP file for Windows - you're looking for the `package` target - but I'm not sure how complete it is, and if you cross-compile it on Linux then you'll end up with LF's in the text files because I need to figure out how to run them through a script.

fabiangreffrath commented on Source/d_main.c:
```diff
@@ -37,6 +37,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 
+#include "config.h"
```
This should already include `config.h` by including `doomdef.h`.

fabiangreffrath commented on Source/i_video.c:
```diff
@@ -28,6 +28,7 @@
 
 #include "SDL.h" // haleyjd
 
+#include "config.h"
```
Same here. `i_video.c` already includes `doomstat.h` which by some other dependency chain already includes `config.h`. BTW, do we define `HAVE_CONFIG_H` anywhere in the CMake build?

AlexMax replied:
Rats, I did not notice that define.  That would explain the errors I was running into.

There are two ways forward.  I can either hard-code that define as a compiler flag, or we can get rid of the define completely, since CMake stamps out `config.h` in all cases.

I would prefer the latter, as it simplifies things in the absence of alternative build systems, but the choice is ultimately yours.

fabiangreffrath commented:
Cool, thanks, please let me know when you consider this ready to merge.

AlexMax commented:
Okay, I'm liking where this is right now.  There is probably more I could do with it, but it should be a capable substitute for the original automake build.  Here are my remaining questions:

- `HAVE_CONFIG_H` - how would you like that to be handled?  Should I hardcode a define at the compiler level, or can I remove the define completely since CMake will always generate a `config.h` file regardless of platform.
- What else do you want in the package?  Right now I have the `COPYING` and `README` files as well as the DLL files on Windows, but it seems like other things could be added, like some of the WAD files.
- Should the app adhere to the FHS on *NIX platforms?  Right now, `make install` puts everything in one directory.  I'm not sure if you want this app to behave like a traditional *NIX app or an `/opt` app.

fabiangreffrath commented:
> * `HAVE_CONFIG_H` - how would you like that to be handled?  Should I hardcode a define at the compiler level, or can I remove the define completely since CMake will always generate a `config.h` file regardless of platform.

If CMake does definitely always produce a `config.h` file, then let's just include it unconditionally. However, I would like to make it clear in the source files that this is a special header and prevent it from getting drowned in the other includes. I'd suggest to include it between the `#include <..>` and `#include "..."` lines with one blank line before and after.

We *should* get away with one `#include "config.h"` in `doomtype.h`, `doomdef.h`, `info.h`, `sounds.h`  and `i_savepng.h` (special case for the check for `dlopen()`), respectively, but maybe you would like to check this again.

> * What else do you want in the package?  Right now I have the `COPYING` and `README` files as well as the DLL files on Windows, but it seems like other things could be added, like some of the WAD files.

I want *all* files to appear in the distribution tarball (woof-x.y.tar.gz).

Then, I would like these files installed as additional documentation:
```
dist_doc_DATA = \
	boom.txt \
	boomdeh.txt \
	boomref.txt \
	dckboom.txt \
	mbf.txt \
	mbfedit.txt \
	mbffaq.txt \
	options.txt \
	winmbf02.txt
```

I would like these files to appear in an `examples` subdirectory:
```
dist_example_DATA = \
	battle.wad \
	dogfly.deh \
	donut.txt \
	donut.wad \
	fireplas.deh \
	fly.deh \
	friend.deh \
	grenade.deh \
	hockey.wad \
	mbfedit.txt \
	mbfedit.wad \
	mine.deh \
	mushroom.deh \
	playbud.deh \
	possbud.deh \
	sky.wad \
	touchy.deh

```

On Unix, I would like the PNG icon and the desktop file to be installed as well.

> * Should the app adhere to the FHS on *NIX platforms?  Right now, `make install` puts everything in one directory.  I'm not sure if you want this app to set up like a traditional *NIX app or an `/opt` app.

Let's please (try to) be FHS compliant. That is, the binary goes to `$(prefix)/bin`, the docs (and the examples subdir) go to `$(prefix)/share/doc/woof`, the icon goes to `$(prefix)/share/icons/hicolor/128x128/apps` and the desktop file to `$(prefix)/share/applications`.

That's how things are currently defined in the `*/Makefile.am` files.

fabiangreffrath commented:
Like this:
```
$ find woof-install/
woof-install/
woof-install/usr
woof-install/usr/bin
woof-install/usr/bin/woof.exe
woof-install/usr/share
woof-install/usr/share/applications
woof-install/usr/share/applications/woof.desktop
woof-install/usr/share/doc
woof-install/usr/share/doc/woof
woof-install/usr/share/doc/woof/boom.txt
woof-install/usr/share/doc/woof/boomdeh.txt
woof-install/usr/share/doc/woof/boomref.txt
woof-install/usr/share/doc/woof/COPYING
woof-install/usr/share/doc/woof/dckboom.txt
woof-install/usr/share/doc/woof/examples
woof-install/usr/share/doc/woof/examples/battle.wad
woof-install/usr/share/doc/woof/examples/dogfly.deh
woof-install/usr/share/doc/woof/examples/donut.txt
woof-install/usr/share/doc/woof/examples/donut.wad
woof-install/usr/share/doc/woof/examples/fireplas.deh
woof-install/usr/share/doc/woof/examples/fly.deh
woof-install/usr/share/doc/woof/examples/friend.deh
woof-install/usr/share/doc/woof/examples/grenade.deh
woof-install/usr/share/doc/woof/examples/hockey.wad
woof-install/usr/share/doc/woof/examples/mbfedit.txt
woof-install/usr/share/doc/woof/examples/mbfedit.wad
woof-install/usr/share/doc/woof/examples/mine.deh
woof-install/usr/share/doc/woof/examples/mushroom.deh
woof-install/usr/share/doc/woof/examples/playbud.deh
woof-install/usr/share/doc/woof/examples/possbud.deh
woof-install/usr/share/doc/woof/examples/sky.wad
woof-install/usr/share/doc/woof/examples/touchy.deh
woof-install/usr/share/doc/woof/mbf.txt
woof-install/usr/share/doc/woof/mbfedit.txt
woof-install/usr/share/doc/woof/mbffaq.txt
woof-install/usr/share/doc/woof/options.txt
woof-install/usr/share/doc/woof/winmbf02.txt
woof-install/usr/share/icons
woof-install/usr/share/icons/hicolor
woof-install/usr/share/icons/hicolor/128x128
woof-install/usr/share/icons/hicolor/128x128/apps
woof-install/usr/share/icons/hicolor/128x128/apps/woof.png
woof-install/usr/share/woof
woof-install/usr/share/woof/betagrph.wad
woof-install/usr/share/woof/betalevl.wad
woof-install/usr/share/woof/watermap.wad
```

AlexMax commented:
I'm working on the commit still, but here's what `ninja package` gives you as of this moment.

```
Woof-0.9.0-Linux/bin/
Woof-0.9.0-Linux/bin/woof
Woof-0.9.0-Linux/share/
Woof-0.9.0-Linux/share/woof/
Woof-0.9.0-Linux/share/woof/betalevl.wad
Woof-0.9.0-Linux/share/woof/watermap.wad
Woof-0.9.0-Linux/share/woof/betagrph.wad
Woof-0.9.0-Linux/share/applications/
Woof-0.9.0-Linux/share/applications/woof.desktop
Woof-0.9.0-Linux/share/doc/
Woof-0.9.0-Linux/share/doc/woof/
Woof-0.9.0-Linux/share/doc/woof/boomref.txt
Woof-0.9.0-Linux/share/doc/woof/winmbf02.txt
Woof-0.9.0-Linux/share/doc/woof/boomdeh.txt
Woof-0.9.0-Linux/share/doc/woof/COPYING
Woof-0.9.0-Linux/share/doc/woof/dckboom.txt
Woof-0.9.0-Linux/share/doc/woof/mbf.txt
Woof-0.9.0-Linux/share/doc/woof/winmbf02s.txt
Woof-0.9.0-Linux/share/doc/woof/README.md
Woof-0.9.0-Linux/share/doc/woof/boom.txt
Woof-0.9.0-Linux/share/doc/woof/options.txt
Woof-0.9.0-Linux/share/doc/woof/mbfedit.txt
Woof-0.9.0-Linux/share/doc/woof/examples/
Woof-0.9.0-Linux/share/doc/woof/examples/dogfly.deh
Woof-0.9.0-Linux/share/doc/woof/examples/donut.txt
Woof-0.9.0-Linux/share/doc/woof/examples/mine.deh
Woof-0.9.0-Linux/share/doc/woof/examples/hockey.wad
Woof-0.9.0-Linux/share/doc/woof/examples/donut.wad
Woof-0.9.0-Linux/share/doc/woof/examples/mbfedit.wad
Woof-0.9.0-Linux/share/doc/woof/examples/mushroom.deh
Woof-0.9.0-Linux/share/doc/woof/examples/fly.deh
Woof-0.9.0-Linux/share/doc/woof/examples/grenade.deh
Woof-0.9.0-Linux/share/doc/woof/examples/battle.wad
Woof-0.9.0-Linux/share/doc/woof/examples/playbud.deh
Woof-0.9.0-Linux/share/doc/woof/examples/mbfedit.txt
Woof-0.9.0-Linux/share/doc/woof/examples/sky.wad
Woof-0.9.0-Linux/share/doc/woof/examples/touchy.deh
Woof-0.9.0-Linux/share/doc/woof/examples/fireplas.deh
Woof-0.9.0-Linux/share/doc/woof/examples/friend.deh
Woof-0.9.0-Linux/share/doc/woof/examples/possbud.deh
Woof-0.9.0-Linux/share/doc/woof/mbf-bugs.html
Woof-0.9.0-Linux/share/doc/woof/mbffaq.txt
Woof-0.9.0-Linux/share/icons/
Woof-0.9.0-Linux/share/icons/hicolor/
Woof-0.9.0-Linux/share/icons/hicolor/128x128/
Woof-0.9.0-Linux/share/icons/hicolor/128x128/apps/
Woof-0.9.0-Linux/share/icons/hicolor/128x128/apps/woof.png
```

And the Windows package, as cross-compiled on Linux...

```
Woof-0.9.0-win64/betalevl.wad
Woof-0.9.0-win64/watermap.wad
Woof-0.9.0-win64/docs/
Woof-0.9.0-win64/docs/boomref.txt
Woof-0.9.0-win64/docs/winmbf02.txt
Woof-0.9.0-win64/docs/boomdeh.txt
Woof-0.9.0-win64/docs/dckboom.txt
Woof-0.9.0-win64/docs/mbf.txt
Woof-0.9.0-win64/docs/winmbf02s.txt
Woof-0.9.0-win64/docs/boom.txt
Woof-0.9.0-win64/docs/options.txt
Woof-0.9.0-win64/docs/mbfedit.txt
Woof-0.9.0-win64/docs/examples/
Woof-0.9.0-win64/docs/examples/dogfly.deh
Woof-0.9.0-win64/docs/examples/donut.txt
Woof-0.9.0-win64/docs/examples/mine.deh
Woof-0.9.0-win64/docs/examples/hockey.wad
Woof-0.9.0-win64/docs/examples/donut.wad
Woof-0.9.0-win64/docs/examples/mbfedit.wad
Woof-0.9.0-win64/docs/examples/mushroom.deh
Woof-0.9.0-win64/docs/examples/fly.deh
Woof-0.9.0-win64/docs/examples/grenade.deh
Woof-0.9.0-win64/docs/examples/battle.wad
Woof-0.9.0-win64/docs/examples/playbud.deh
Woof-0.9.0-win64/docs/examples/mbfedit.txt
Woof-0.9.0-win64/docs/examples/sky.wad
Woof-0.9.0-win64/docs/examples/touchy.deh
Woof-0.9.0-win64/docs/examples/fireplas.deh
Woof-0.9.0-win64/docs/examples/friend.deh
Woof-0.9.0-win64/docs/examples/possbud.deh
Woof-0.9.0-win64/docs/mbf-bugs.html
Woof-0.9.0-win64/docs/mbffaq.txt
Woof-0.9.0-win64/libvorbisfile-3.dll
Woof-0.9.0-win64/libogg-0.dll
Woof-0.9.0-win64/SDL2_net.dll
Woof-0.9.0-win64/libmodplug-1.dll
Woof-0.9.0-win64/SDL2.dll
Woof-0.9.0-win64/libvorbis-0.dll
Woof-0.9.0-win64/betagrph.wad
Woof-0.9.0-win64/libFLAC-8.dll
Woof-0.9.0-win64/libopus-0.dll
Woof-0.9.0-win64/COPYING.txt
Woof-0.9.0-win64/SDL2_mixer.dll
Woof-0.9.0-win64/libmpg123-0.dll
Woof-0.9.0-win64/libopusfile-0.dll
Woof-0.9.0-win64/README.txt
Woof-0.9.0-win64/woof.exe
```

chungy commented:
> I'm working on the commit still, but here's what `ninja package` gives me as of this moment.
> 
> ```
> Woof-0.9.0-Linux/bin/
> Woof-0.9.0-Linux/bin/woof
> ```

It's fairly strange in the Linux world to unpack an application to run like this (it's certainly possible, but strange). If you can integrate AppImage generation instead (single-file executable apps), that would be a lot better.

AlexMax commented:
> It's fairly strange in the Linux world to unpack an application to run like this (it's certainly possible, but strange). If you can integrate AppImage generation instead (single-file executable apps), that would be a lot better.

`package` is used for distributing binaries.  There is also a `package_source` target that bundles the source tree.

fabiangreffrath commented:
I am about to merge this soon (will unconditionally enable BETA and DOGS before). Please put a license header into all files which you consider under your copyright.

AlexMax commented:
Do I need to put a copyright header into every build file?  I find it a little strange, since I so rarely see copyright notices in build files, including the existing automake files for both this project and Chocolate Doom.

I think I did end up adding licenses to the files that I might reasonably expect to be reused between projects, which was the `win-textfile` Python script and the FindSDL cmake files that I've been copy-pasting and modifying between projects.

fabiangreffrath commented:
That's fine with me, thanks!

chungy commented:
> I think I did end up adding licenses to the files that I might reasonably expect to be reused between projects, which was the win-textfile Python script and the FindSDL cmake files that I've been copy-pasting and modifying between projects.

Hmm, I do have two questions about these:

1. Why not just depend on unix2dos? Part of the [dos2unix](https://waterlan.home.xs4all.nl/dos2unix.html) project and available in basically every distro repo
2. Shouldn't you use the FindSDL.cmake file installed on the system rather than a local copy?

chungy commented:
There's also a heck of a lot more DLL files included compared to the automake version of win32 packaging -- 28 versus 5. Maybe not a real problem but it looks messier :P

AlexMax commented:
> Why not just depend on unix2dos? Part of the dos2unix project and available in basically every distro repo.

I was attempting to avoid another dependency.

> Shouldn't you use the FindSDL.cmake file installed on the system rather than a local copy?

My finders are superior in many ways to the one included with CMake.  It includes the `SDL2_*_DIR` variables that are very handy for pointing at the root of an SDL development library on Windows, for ease of use.  It also exposes the library and its dependent compiler flags as imported targets, which makes it much easier to link against.  It also properly uses pkg-config to find SDL2 installed from a distribution.

>  OK I used i686-w64-mingw32-cmake to see how well it works, seems to be mostly fine, though related to point 1, you're modifying the in-repo text files directly instead of copying to a new location. It's better if you copy-and-modify instead and then git status doesn't show a bunch of modified text files.

I have developed this build system with the intention of doing out-of-tree builds.  I went ahead and added a conditional that prevents it outright.

chungy commented:
> I have developed this build system with the intention of doing out-of-tree builds. I went ahead and added a conditional that prevents in-tree builds outright. You'll notice in the gitignore that I added an ignore for any directory that begins with the word build as that seems to be the standard.

Thanks, that works (though it's the kind of complicated thing that turns me off of CMake, honestly...)

> ```
> % cmake .. -G Ninja -DCMAKE_BUILD_TYPE=Debug \
>     -DCMAKE_TOOLCHAIN_FILE=../cmake/CrossToWin64.cmake \
>     -DSDL2_DIR=../../mingw/SDL2-2.0.10 \
>     -DSDL2_MIXER_DIR=../../mingw/SDL2_mixer-2.0.4/ \
>     -DSDL2_NET_DIR=../../mingw/SDL2_net-2.0.1
> ```

I wasn't really sure what values to put in for the *DIR stuff, but I placed `/usr/i686-w64-mingw32` there anyway and it seemed to work. Overall it seems easier to just keep using the [i686-w64-mingw32-cmake](https://aur.archlinux.org/cgit/aur.git/tree/mingw-cmake.sh?h=mingw-w64-cmake) script to do this :P

AlexMax commented:
> I wasn't really sure what values to put in for the *DIR stuff

On windows, or when you're cross-compiling using mingw, you point those at the root of the extracted [development library](https://www.libsdl.org/download-2.0.php).  In the past, I've also had success pointing those variables at something like `/usr/local` for times when i was compiling with a custom SDL.

If CMake finds the appropriate version of SDL out of the box, as it should on Linux with distro-supplied versions of SDL, you don't need to touch those cache variables at all.

chungy commented:
> On windows, or when you're cross-compiling using mingw, you point those at the root of the extracted development library.

Yeah I've been testing with cross-compiling, though I don't manually extract anything, it's just a package-installed mingw-w64-sdl2 (and related)

AlexMax commented:
> There's also a heck of a lot more DLL files included compared to the automake version of win32 packaging -- 28 versus 5. Maybe not a real problem but it looks messier :P

Where are you getting 28 from?  That's an assumption I made based on the downloadable development library directly from SDL - if you're installing it directly from package repositories that might be a problem that I need to think of a fix for.

chungy commented:
> Where are you getting 28 from?

From the result of `make package`

AlexMax commented:
I meant "Where did the 28 DLL's come from?"  Do they look like SDL and related libraries, because I suspect that because you installed mingw SDL from repositories, it might be grabbing the entire contents of the mingw lib directory.

chungy commented:
Oh yeah, that seems to be it.

```
$ ls /usr/i686-w64-mingw32/bin/*.dll | wc -l
28
```

AlexMax commented:
Dang.  I will need some thought on how to solve this in a way that preserves both workflows.
